packages/cli/commands/project/watch.js

```diff
@@ -0,0 +1,268 @@
+};
+
+const currentBuild = {
+  id: null,
+  isFetchingNewBuildId: false,
+  get: () => {
+    return this.id;
+  },
+  update: async (accountId, projectName) => {
+    if (this.id) {
+      return this.id;
+    }
+    if (this.isFetchingNewBuildId) {
+      return;
+    }
+    logger.log(i18n(`${i18nKey}.logs.createNewBuild`));
+    this.isFetchingNewBuildId = true;
+    this.id = await createNewBuild(accountId, projectName);
+    this.isFetchingNewBuildId = false;
+  },
+  clear: () => {
+    this.id = null;
+  },
+};
+
+const debounceQueueBuild = (accountId, projectName) => {
+  if (timer) {
+    clearTimeout(timer);
+  }
+
+  timer = setTimeout(async () => {
+    logger.debug(i18n(`${i18nKey}.debug.pause`, { projectName }));
+    queue.pause();
+    await queue.onIdle();
+
+    try {
+      await queueBuild(accountId, projectName, currentBuild.get());
+      logger.debug(i18n(`${i18nKey}.debug.buildStarted`, { projectName }));
+    } catch (err) {
+      logApiErrorInstance(err, new ApiErrorContext({ accountId, projectName }));
+      return;
+    }
+
+    await pollBuildStatus(accountId, projectName, currentBuild.get());
+    currentBuild.clear();
+
+    if (standbyeQueue.length > 0) {
+      await processStandByQueue(accountId, projectName);
+    }
+
+    queue.start();
+    logger.log(i18n(`${i18nKey}.logs.resuming`));
+  }, 5000);
+};
+
+const queueFileUpload = async (
+  accountId,
+  projectName,
+  filePath,
+  remotePath
+) => {
+  if (!isAllowedExtension(filePath)) {
+    logger.debug(i18n(`${i18nKey}.debug.extensionNotAllowed`, { filePath }));
+    return;
+  }
+  if (shouldIgnoreFile(filePath)) {
+    logger.debug(i18n(`${i18nKey}.debug.ignored`, { filePath }));
+    return;
+  }
+  if (!queue.isPaused) {
+    debounceQueueBuild(accountId, projectName);
+  }
+
+  logger.debug(i18n(`${i18nKey}.debug.uploading`, { filePath, remotePath }));
+
+  return queue.add(async () => {
+    await currentBuild.update(accountId, projectName);
+    try {
+      await uploadFileToBuild(
+        accountId,
+        projectName,
+        currentBuild.get(),
+        filePath,
+        remotePath
+      );
+      logger.log(
+        i18n(`${i18nKey}.logs.uploadSucceeded`, { filePath, remotePath })
+      );
+    } catch (err) {
+      logger.debug(
+        i18n(`${i18nKey}.debug.uploadFailed`, { filePath, remotePath })
+      );
+    }
+  });
+};
+
+const createNewBuild = async (accountId, projectName) => {
+  try {
+    logger.debug('Attempting to create a new build');
+    const { buildId } = await provisionBuild(accountId, projectName);
+    return buildId;
+  } catch (err) {
+    if (err.error.subCategory === 'PipelineErrors.PROJECT_LOCKED') {
+      logger.error('Project is locked, cannot create new build');
+    } else if (err.error.subCategory === 'PipelineErrors.MISSING_PROJECT') {
+      logger.error(`Project ${projectName} does not exist.`);
+    } else {
+      logApiErrorInstance(err, new ApiErrorContext({ accountId, projectName }));
```

Comment:
If we get here, I think we should cancel the build so that it ends up being unlocked.